### PR TITLE
Changed how relay loops are started

### PR DESCRIPTION
**Pull request by svyatonik**
This is required both for #833 (otherwise process with 4 relays will provide 4 same global metrics, spending 4x time on sysinfo calls) and for "standalone" metrics (see https://github.com/paritytech/parity-bridges-common/pull/822#discussion_r594091027).

### CODE REVIEW

File: relays/generic/exchange/src/exchange_loop.rs [hunk 1]
```diff
@@ -85,50 +85,32 @@ pub async fn run<P: TransactionProofPipeline>(
...
-		source_client,
-		target_client,
-		|source_client, target_client| {
+
+	relay_utils::relay_loop(source_client, target_client)
+		.with_metrics(format!("{}_to_{}_Exchange", P::SOURCE_NAME, P::TARGET_NAME))
+		.loop_metric(ExchangeLoopMetrics::default())?
+		.standalone_metric(GlobalMetrics::default())?
+		.expose(metrics_params)
+		.await?
```
Comment (svyatonik) [inline relays/generic/exchange/src/exchange_loop.rs:96]: I'd prefer to have additional tabs here and in other loop files, but fmt prevents me from doing this :/

File: relays/generic/utils/src/metrics/global.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
```
Comment (svyatonik) [inline relays/generic/utils/src/metrics/global.rs:1]: There's now single file in this folder, but I have another one to add in the next PRs

File: relays/utils/src/relay_loop.rs [hunk 3]
```diff
@@ -32,60 +34,164 @@ pub trait Client: Clone + Send + Sync {
...
+			loop_metric: None,
+		}
+	}
+
+	/// Run relay loop.
+	///
+	/// This function represents an outer loop, which in turn calls provided `loop_run` function to do
+	/// actual job. When `loop_run` returns, this outer loop reconnects to failed client (source,
+	/// target or both) and calls `loop_run` again.
+	pub async fn run<R, F>(mut self, loop_run: R) -> Result<(), String>
```
Comment (HCastano) [inline relays/utils/src/relay_loop.rs:84]:
```suggestion
	pub async fn run<R, F>(mut self, run_loop: R) -> Result<(), String>
```

Comment (HCastano) [reply relays/utils/src/relay_loop.rs:84]: From what I understand this doesn't need to be any sort of loop though, it really just needs to be something which returns a `Future`, right? Maybe we shouldn't even have loop in the name here

File: relays/utils/src/relay_loop.rs [hunk 4]
```diff
@@ -32,60 +34,164 @@ pub trait Client: Clone + Send + Sync {
...
+
+		Ok(())
+	}
+}
+
+impl<SC, TC, LM> LoopMetrics<SC, TC, LM> {
+	/// Add relay loop metrics.
+	///
+	/// Loop metrics will be passed to the loop callback.
+	pub fn loop_metric<NewLM: Metrics>(self, loop_metric: NewLM) -> Result<LoopMetrics<SC, TC, NewLM>, String> {
```
Comment (HCastano) [inline relays/utils/src/relay_loop.rs:148]:
```suggestion
	pub fn loop_metric<M: Metrics>(self, loop_metric: M) -> Result<LoopMetrics<SC, TC, M>, String> {
```

File: relays/utils/src/metrics.rs [hunk 5]
```diff
@@ -31,62 +32,32 @@ pub struct MetricsParams {
 }
 
 /// Metrics API.
-pub trait Metrics {
+pub trait Metrics: Clone + Send + Sync + 'static {
```
Comment (HCastano) [inline relays/utils/src/metrics.rs:35]: Why does this need to be `'static` now?

**Review by HCastano — APPROVED** [review]
>  I'd prefer to have additional tabs here and in other loop files, but fmt prevents me from doing this :/

I would say that RustFmt's word is law, even if we don't agree with it.

Also, looks like there are some Clippy warnings which need to be fixed

File: relays/utils/src/metrics.rs (hunk 5 above)
Comment (svyatonik) [reply relays/utils/src/metrics.rs:35]: It is now used in [`async_std::task::spawn()`](https://docs.rs/async-std/1.9.0/async_std/task/fn.spawn.html), which required future to be `'static`

File: relays/utils/src/relay_loop.rs (hunk 3 above)
Comment (svyatonik) [reply relays/utils/src/relay_loop.rs:84]: Yes, you may use it to run arbitrary futures, I see no point in using abstract names here. It is only used (and most probably will only be used) to run relay loops. I bet you'll be the first one to object if I'll rename this whole functionality to something like `fn run_future_with_retry()` :)

**Pull request merged**